File: DSA/src/Questions/BinarySearch/RandomPickWeight.java
package Questions.BinarySearch;

import java.util.Random;
/*
https://leetcode.com/problems/random-pick-with-weight/description/
 You are given a 0-indexed array of positive integers w where w[i] describes the weight of the ith index.

You need to implement the function pickIndex(), which randomly picks an index in the range [0, w.length - 1] (inclusive) and returns it. The probability of picking an index i is w[i] / sum(w).

For example, if w = [1, 3], the probability of picking index 0 is 1 / (1 + 3) = 0.25 (i.e., 25%), and the probability of picking index 1 is 3 / (1 + 3) = 0.75 (i.e., 75%).
 

Example 1:

Input
["Solution","pickIndex"]
[[[1]],[]]
Output
[null,0]

Explanation
Solution solution = new Solution([1]);
solution.pickIndex(); // return 0. The only option is to return 0 since there is only one element in w.
Example 2:

Input
["Solution","pickIndex","pickIndex","pickIndex","pickIndex","pickIndex"]
[[[1,3]],[],[],[],[],[]]
Output
[null,1,1,1,1,0]

Explanation
Solution solution = new Solution([1, 3]);
solution.pickIndex(); // return 1. It is returning the second element (index = 1) that has a probability of 3/4.
solution.pickIndex(); // return 1
solution.pickIndex(); // return 1
solution.pickIndex(); // return 1
solution.pickIndex(); // return 0. It is returning the first element (index = 0) that has a probability of 1/4.

Since this is a randomization problem, multiple answers are allowed.
All of the following outputs can be considered correct:
[null,1,1,1,1,0]
[null,1,1,1,1,1]
[null,1,1,1,0,0]
[null,1,1,1,0,1]
[null,1,0,1,0,0]
......
and so on.
Time Complexity: O(logN)
Space Complexity: O(N)
 */
public class RandomPickWeight {
 Random random;
    int[] wSums;

    public RandomPickWeight(int[] w) {
        this.random = new Random();
        for (int i = 1; i < w.length; ++i)
            w[i] += w[i - 1];
        this.wSums = w;
    }

    public int pickIndex() {
        int len = wSums.length;
        int idx = random.nextInt(wSums[len - 1]) + 1;
        int left = 0, right = len - 1;
        // search position
        while (left < right) {
            int mid = (right + left) >> 1;
            if (wSums[mid] == idx)
                return mid;
            else if (wSums[mid] < idx)
                left = mid + 1;
            else
                right = mid;
        }
        return left;
    }
    public static void main(String[] args) {
        int arr[] = { 3, 34, 4, 12, 5, 2 };
        RandomPickWeight randomPickWeight = new RandomPickWeight(arr);
        System.out.println(randomPickWeight.pickIndex());
    }
}

File: DSA/Python/Questions/BinarySearch/RandomPickWeight.py
"""
# Question: Random Pick with Weight
# Link: https://leetcode.com/problems/random-pick-with-weight/

# Time Complexity: 
# - Constructor: O(n)
# - PickIndex: O(log n)
# Space Complexity: O(n)

# Algorithm:
# 1. Create cumulative sum array
# 2. Binary search for random value
# 3. Return weighted random index
"""

import random


class RandomPickWeight:
    def __init__(self, w: list[int]):
        self.prefix_sums = []
        prefix_sum = 0
        for weight in w:
            prefix_sum += weight
            self.prefix_sums.append(prefix_sum)
        self.total_sum = prefix_sum

    def pick_index(self) -> int:
        target = random.random() * self.total_sum
        left, right = 0, len(self.prefix_sums) - 1

        while left < right:
            mid = (left + right) >> 1
            if self.prefix_sums[mid] > target:
                right = mid
            else:
                left = mid + 1

        return left


def main():
    solution = RandomPickWeight([1, 3, 2, 4])
    print(solution.pick_index())  # Random output based on weights


if __name__ == "__main__":
    main()
